File: src/main/java/com/itranswarp/summer/context/ApplicationContext.java
package com.itranswarp.summer.context;

import org.jetbrains.annotations.Nullable;

import java.util.List;


/**
 * Used for BeanPostProcessor.
 */
public interface ApplicationContext extends AutoCloseable {
    /**
     * 是否存在指定name的Bean？
     */
    boolean containsBean(String name);

    /**
     * 根据name返回唯一Bean，未找到抛出NoSuchBeanDefinitionException
     */
    <T> T getBean(String name);

    /**
     * 根据name返回唯一Bean，未找到抛出NoSuchBeanDefinitionException，找到但type不符抛出BeanNotOfRequiredTypeException
     */
    <T> T getBean(String name, Class<T> requiredType);

    /**
     * 根据type返回唯一Bean，未找到抛出NoSuchBeanDefinitionException
     */
    <T> T getBean(Class<T> requiredType);

    /**
     * 根据type返回一组Bean，未找到返回空List
     */
    <T> List<T> getBeans(Class<T> requiredType);

    /**
     * 关闭并执行所有bean的destroy方法
     */
    void close();


    List<BeanDefinition> findBeanDefinitions(Class<?> type);

    @Nullable
    BeanDefinition findBeanDefinition(Class<?> type);

    @Nullable
    BeanDefinition findBeanDefinition(String name);

    @Nullable
    BeanDefinition findBeanDefinition(String name, Class<?> requiredType);

    Object createBeanAsEarlySingleton(BeanDefinition def);
}

File: src/main/java/com/itranswarp/summer/ApplicationContext.kt
package com.itranswarp.summer

/**
 * Used for BeanPostProcessor.
 */
interface ApplicationContext : AutoCloseable, Cloneable {
    /**
     * Is the bean existed?
     */
    fun containsBean(name: String): Boolean

    /**
     * Return the unique bean by name
     * if it does not exist, will throw NoSuchBeanDefinitionException
     */
    fun <T> getBean(name: String): T

    /**
     * Return the unique bean by name
     * if it does not exist, will throw NoSuchBeanDefinitionException
     * if bean is not required type, will throw BeanNotOfRequiredTypeException
     */
    fun <T> getBean(name: String, requiredType: Class<T>): T

    /**
     * Return the unique bean by type
     * if it does not exist, will throw NoSuchBeanDefinitionException
     * @param requiredType the type of beans
     * @return the unique bean
     */
    fun <T> getBean(requiredType: Class<T>): T

    /**
     * Return a group of bean by type
     * if it does not exist, will return an empty list
     * @param requiredType the type of beans
     * @return List of beans
     */
    fun <T> getBeans(requiredType: Class<T>): List<T>?

    /**
     * Close the context and call beans' destroy method
     */
    override fun close()
    fun findBeanDefinitions(type: Class<*>): List<BeanDefinition?>?
    fun findBeanDefinition(type: Class<*>): BeanDefinition?
    fun findBeanDefinition(name: String): BeanDefinition?
    fun findBeanDefinition(name: String, requiredType: Class<*>?): BeanDefinition?
    fun createBeanAsEarlySingleton(def: BeanDefinition): Any?
    val beanDefinitions: Collection<BeanDefinition?>?
}
